pkg/mutation/mutation_function_test.go

```diff
@@ -0,0 +1,325 @@
+				},
+				{
+					Name:  "testname2",
+					Image: "image",
+					Ports: []corev1.ContainerPort{
+						{Name: "portName2A"},
+						{Name: "portName2B"},
+					},
+				},
+				{
+					Name:  "testname3",
+					Image: "image",
+					Ports: []corev1.ContainerPort{{Name: "portName3"}},
+				},
+			},
+		},
+	}
+	podObject, err := runtime.DefaultUnstructuredConverter.ToUnstructured(pod)
+	if err != nil {
+		t.Errorf("Failed to convert pod to unstructured %v", err)
+	}
+	return &unstructured.Unstructured{Object: podObject}
+}
+
+func TestObjects(t *testing.T) {
+	testFunc := func(unstr *unstructured.Unstructured) {
+		element, found, err := unstructured.NestedString(unstr.Object, "metadata", "labels", "labelA")
+		if err != nil {
+			t.Error("Unexpected error", err)
+		}
+		if !found && element != TestValue {
+			t.Errorf("Failed to update pod")
+		}
+	}
+	testAssignMetadataMutation(
+		"metadata.labels.labelA",
+		TestValue,
+		prepareTestPod(t),
+		testFunc,
+		t,
+	)
+}
+
+func TestObjectsAndLists(t *testing.T) {
+	testFunc := func(u *unstructured.Unstructured) {
+		containers, _, err := unstructured.NestedFieldNoCopy(u.Object, "spec", "containers")
+		if err != nil {
+			t.Error("Unexpected error", err)
+		}
+		for _, container := range containers.([]interface{}) {
+			containerAsMap := container.(map[string]interface{})
+			if containerAsMap["name"] == "testname2" {
+				for _, port := range containerAsMap["ports"].([]interface{}) {
+					portAsMap := port.(map[string]interface{})
+					if portAsMap["name"] == "portName2B" {
+						if portAsMap["hostIP"] != TestValue {
+							t.Errorf("Failed to update pod")
+						}
+					} else {
+						if _, ok := port.(map[string]interface{})["hostIP"]; ok {
+							t.Errorf("Unexpected pod was updated")
+						}
+					}
+				}
+			} else {
+				for _, port := range container.(map[string]interface{})["ports"].([]interface{}) {
+					if _, ok := port.(map[string]interface{})["hostIP"]; ok {
+						t.Errorf("Unexpected pod was updated")
+					}
+				}
+			}
+		}
+	}
+
+	testAssignMutation(
+		`spec.containers["name": "testname2"].ports["name": "portName2B"].hostIP`,
+		TestValue,
+		prepareTestPod(t),
+		testFunc,
+		t,
+	)
+}
+
+func TestListsAsLastElement(t *testing.T) {
+	testFunc := func(u *unstructured.Unstructured) {
+		containers, _, err := unstructured.NestedFieldNoCopy(u.Object, "spec", "containers")
+		if err != nil {
+			t.Error("Unexpected error", err)
+		}
+		for _, container := range containers.([]interface{}) {
+			if container.(map[string]interface{})["name"] == TestValue {
+				return
+			}
+		}
+		t.Errorf("Failed to update pod")
+	}
+
+	testAssignMutation(
+		`spec.containers["name": "notExists"]`,
+		TestValue,
+		prepareTestPod(t),
+		testFunc,
+		t,
+	)
+}
+
+func TestListsAsLastElementAlreadyExists(t *testing.T) {
+	testFunc := func(u *unstructured.Unstructured) {
+		containers, _, err := unstructured.NestedFieldNoCopy(u.Object, "spec", "containers")
```

Comment:
`unstructured.NestedSlice` will take care of the typecasting for you here.

Reply:
Done